File: C/churn.c
#define LINUX 1
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <dirent.h>

#ifdef __GNUC__

#include <sys/stat.h>
#include <sys/types.h>
#define FILE_INFO                 struct dirent
#define FIND_FIRST( dir, info )   ( ( info = readdir( dir ) ) != NULL )
#define FILE_IS_DIR( path, info ) ( stat( path, &st ) == 0 && S_ISDIR( st.st_mode ) )
#define FILE_NAME( info )         ( ( info )->d_name )

#else

#error "Unsupported compiler."

#endif

int total_files;
int total_passed;
int total_failed;
char *compress_command;
char *expand_command;
FILE *input;
FILE *output;
FILE *compressed;
FILE *log_file;

void churn_files( const char *path );
int file_is_already_compressed( const char *name );
void close_all_the_files( void );
int compress( const char *file_name );
void usage_exit( void );

void main( int argc, char *argv[] )
{
    time_t start_time;
    time_t stop_time;
    char root_dir[ 1024 ];

    if ( argc != 4 )
        usage_exit();
    strcpy( root_dir, argv[ 1 ] );
    if ( root_dir[ strlen( root_dir ) - 1 ] != '/' )
       strcat( root_dir, "/" );
    compress_command = argv[ 2 ];
    expand_command = argv[ 3 ];
    setbuf( stdout, NULL );
    setbuf( stderr, NULL );
    total_files = 0;
    total_passed = 0;
    total_failed = 0;
    log_file = fopen( "CHURN.LOG", "w" );
    if ( log_file == NULL ) {
        printf( "Couldn't open the log file!\n" );
        exit( 1 );
    }
    fprintf( log_file, "                                          "
            "Original   Packed\n" );
    fprintf( log_file, "            File Name                     "
            "  Size      Size   Ratio  Result\n" );
    fprintf( log_file, "-------------------------------------     "
            "--------  --------  ----  ------\n" );
    time( &start_time );
    churn_files( root_dir );
    time( &stop_time );
    fprintf( log_file, "\nTotal elapsed time: %f seconds\n",
             difftime( stop_time, start_time ) );
    fprintf( log_file, "Total files:   %d\n", total_files );
    fprintf( log_file, "Total passed:  %d\n", total_passed );
    fprintf( log_file, "Total failed:  %d\n", total_failed );
}


void churn_files( const char *path )
{
    DIR *dir;
    FILE_INFO *file_info;
    struct stat st;
    char full_name[ 1024 ];

    dir = opendir( path );
    if ( dir == NULL ) {
        perror( "Could not open directory" );
        exit( 1 );
    }

    while ( FIND_FIRST( dir, file_info ) ) {
        if ( strcmp( FILE_NAME( file_info ), "." ) == 0 || strcmp( FILE_NAME( file_info ), ".." ) == 0 )
            continue;

        snprintf( full_name, sizeof(full_name), "%s/%s", path, FILE_NAME( file_info ) );
        if ( FILE_IS_DIR( full_name, file_info ) ) {
            churn_files( full_name );
        } else {
            if ( !file_is_already_compressed( full_name ) ) {
                fprintf( stderr, "Testing %s\n", full_name );
                if ( !compress( full_name ) )
                    fprintf( stderr, "Comparison failed!\n" );
            }
        }
    }

    closedir( dir );
}


int file_is_already_compressed( const char *name )
{
    const char *extension;
    static const char *matches[]={ "zip", "gz", "bz2", "xz", "tar", "rar", NULL };
    int i;

    extension = strrchr( name, '.' );
    if ( extension == NULL || ++extension == NULL )
        return( 0 );
    i = 0;
    while ( matches[ i ] != NULL )
        if ( strcmp( extension, matches[ i++ ] ) == 0 )
            return( 1 );
    return( 0 );
}


int compress( const char *file_name )
{
    long new_size;
    long old_size;
    int c;
    char command[ 132 ];

    printf( "%s\n", file_name );
    fprintf( log_file, "%-40s ", file_name );
    snprintf( command, sizeof(command), compress_command, file_name );
    system( command );
    snprintf( command, sizeof(command), expand_command, file_name );
    system( command );

    input = fopen( file_name, "rb" );
    output = fopen( "TEST.OUT", "rb" );
    compressed = fopen( "TEST.CMP", "rb" );

    total_files++;
    if ( input == NULL || output == NULL || compressed == NULL ) {
        total_failed++;
        close_all_the_files();
        fprintf( log_file, "Failed, couldn't open file!\n" );
        return( 0 );
    }

    fseek( input, 0L, SEEK_END );
    old_size = ftell( input );
    fseek( input, 0L, SEEK_SET );
    fseek( compressed, 0L, SEEK_END );
    new_size = ftell( compressed );

    fprintf( log_file, " %8ld  %8ld ", old_size, new_size );
    if ( old_size == 0L )
        old_size = 1L;
    fprintf( log_file, "%4ld%%  ",
            100L - ( ( 100L * new_size ) / old_size ) );
    do {
        c = getc( input );
        if ( getc( output ) != c ) {
            fprintf( log_file, "Failed\n" );
            total_failed++;
            close_all_the_files();
            return( 0 );
        }
    }
    while ( c != EOF );
    fprintf( log_file, "Passed\n" );
    close_all_the_files();
    total_passed++;
    return( 1 );
}

void close_all_the_files()
{
    if ( input != NULL )
        fclose( input );
    if ( output != NULL )
        fclose( output );
    if ( compressed != NULL )
        fclose( compressed );
}


void usage_exit( void )
{
    const char *usage = "CHURN 1.0. Usage:  CHURN root-dir \"compress "
                  "command\" \"expand command\n"
                  "\n"
                  "CHURN is used to test compression programs.  "
                  "It does this by compressing and\n"
                  "then expanding all of the files in and under "
                  "the specified root dir.\n"
                  "\n"
                  "For each file it finds, CHURN first executes "
                  "the compress command to create a\n"
                  "compressed file called TEST.CMP.  It then "
                  "executes the expand command to\n"
                  "create a file called TEST.OUT. CHURN then "
                  "compares the two files to make sure\n"
                  "the compression cycle worked properly.\n"
                  "\n"
                  "The file name to be compressed will be "
                  "inserted into the compress command\n"
                  "using sprintf, with any %%s argument being "
                  "substituted with the name of the\n"
                  "file being compressed.  Note that the "
                  "compress and expand commands should be\n"
                  "enclosed in double quotes so that multiple "
                  "words can be included in the\n"
                  "commands.\n"
                  "\n"
                  "Note that you may have to double the %% "
                  "character on your command line to get\n"
                  "around argument substitution under some "
                  "command processors. Finally, note that\n"
                  "CHURN executes the compression program "
                  "using a system() function call, so\n"
                  "batch files can be used to execute complex "
                  "compression sequences.\n"
                  "\n"
                  "Example:  CHURN / \"gzip -c %%s > TEST.CMP\""
                  " \"gzip -d -c TEST.CMP > TEST.OUT\"";

    puts( usage );
    exit( 1 );
}

File: Python/churn.py
import os
import sys
import subprocess
import time

class ChurnProgram:
    def __init__(self):
        self.total_files = 0
        self.total_passed = 0
        self.total_failed = 0
        self.compress_command = ""
        self.expand_command = ""
        self.log_file = None
        
    def main(self, args):
        if len(args) != 3:
            self.usage_exit()
            return
            
        root_dir = args[0]
        self.compress_command = args[1]
        self.expand_command = args[2]
        
        # Ensure trailing separator
        if not root_dir.endswith(os.sep):
            root_dir = root_dir + os.sep
        
        try:
            self.log_file = open("CHURN.LOG", "w", encoding="utf-8")
            self.write_log_header()
            
            start_time = time.time()
            self.churn_files(root_dir)
            stop_time = time.time()
            
            self.write_log_summary(start_time, stop_time)
            self.log_file.close()
            
        except Exception as e:
            print(f"Error: {e}", file=sys.stderr)
            sys.exit(1)
    
    def churn_files(self, path):
        """Recursively process all files in directory"""
        try:
            # Process all files in current directory
            for entry in os.scandir(path):
                if entry.is_dir(follow_symlinks=False):
                    self.churn_files(entry.path)
                elif entry.is_file(follow_symlinks=False):
                    if not self.file_is_already_compressed(entry.path):
                        print(f"Testing {entry.path}", file=sys.stderr)
                        if not self.compress(entry.path):
                            print("Comparison failed!", file=sys.stderr)
                            
        except PermissionError as e:
            print(f"Access denied to {path}: {e}", file=sys.stderr)
        except Exception as e:
            print(f"Error processing {path}: {e}", file=sys.stderr)
    
    def file_is_already_compressed(self, filename):
        """Check if file already has a compressed extension"""
        compressed_extensions = {'.zip', '.ice', '.lzh', '.arc', '.gif', '.pak', '.arj', 
                                 '.gz', '.bz2', '.xz', '.7z', '.rar', '.tar', '.tgz'}
        extension = os.path.splitext(filename)[1].lower()
        return extension in compressed_extensions
    
    def compress(self, filename):
        """Compress and expand a file, then compare results"""
        print(filename)
        
        # Prepare temporary files with absolute paths
        current_dir = os.getcwd()
        compressed_file = os.path.join(current_dir, "TEST.CMP")
        output_file = os.path.join(current_dir, "TEST.OUT")
        
        # Clean up any existing temp files
        for temp_file in [compressed_file, output_file]:
            if os.path.exists(temp_file):
                try:
                    os.remove(temp_file)
                except:
                    pass
        
        try:
            # Log file entry
            self.log_file.write(f"{os.path.basename(filename):<40} ")
            
            # Handle different command formats
            # For compress command, replace %s with the filename
            compress_cmd = self.compress_command
            
            # Check if command uses TEST.CMP as output file
            if "TEST.CMP" in compress_cmd:
                # Replace TEST.CMP with actual path
                compress_cmd = compress_cmd.replace("TEST.CMP", compressed_file)
                # Replace %s with filename
                compress_cmd = compress_cmd.replace("%s", f'"{filename}"')
                
                # Execute compress command
                result = subprocess.run(
                    compress_cmd,
                    shell=True,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    timeout=30
                )
            else:
                # Assume command writes to stdout, redirect to TEST.CMP
                compress_cmd = compress_cmd.replace("%s", f'"{filename}"')
                with open(compressed_file, "wb") as outfile:
                    result = subprocess.run(
                        compress_cmd,
                        shell=True,
                        stdout=outfile,
                        stderr=subprocess.PIPE,
                        timeout=30
                    )
            
            if result.returncode != 0:
                error_msg = result.stderr.decode('utf-8', errors='ignore') if result.stderr else ""
                print(f"Compress command failed: {error_msg}", file=sys.stderr)
                self.log_file.write("Failed: Compress command failed\n")
                self.total_failed += 1
                return False
            
            # Execute expand command
            expand_cmd = self.expand_command
            
            # Check if command uses TEST.CMP as input and TEST.OUT as output
            if "TEST.CMP" in expand_cmd and "TEST.OUT" in expand_cmd:
                # Replace TEST.CMP and TEST.OUT with actual paths
                expand_cmd = expand_cmd.replace("TEST.CMP", compressed_file)
                expand_cmd = expand_cmd.replace("TEST.OUT", output_file)
                
                result = subprocess.run(
                    expand_cmd,
                    shell=True,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                    timeout=30
                )
            elif "TEST.CMP" in expand_cmd:
                # Command reads from TEST.CMP and writes to stdout
                expand_cmd = expand_cmd.replace("TEST.CMP", compressed_file)
                with open(output_file, "wb") as outfile:
                    result = subprocess.run(
                        expand_cmd,
                        shell=True,
                        stdout=outfile,
                        stderr=subprocess.PIPE,
                        timeout=30
                    )
            else:
                # Unknown format
                self.log_file.write("Failed: Invalid expand command format\n")
                self.total_failed += 1
                return False
            
            if result.returncode != 0:
                error_msg = result.stderr.decode('utf-8', errors='ignore') if result.stderr else ""
                print(f"Expand command failed: {error_msg}", file=sys.stderr)
                self.log_file.write("Failed: Expand command failed\n")
                self.total_failed += 1
                return False
            
            # Check if output files were created
            if not os.path.exists(compressed_file):
                self.log_file.write("Failed: Compressed file not created\n")
                self.total_failed += 1
                return False
            
            if not os.path.exists(output_file):
                self.log_file.write("Failed: Output file not created\n")
                self.total_failed += 1
                return False
            
            # Get file sizes
            old_size = os.path.getsize(filename)
            new_size = os.path.getsize(compressed_file)
            
            self.total_files += 1
            
            # Log sizes and ratio
            self.log_file.write(f" {old_size:8} {new_size:8} ")
            if old_size == 0:
                old_size = 1
            
            ratio = 100 - (new_size * 100 // old_size)
            self.log_file.write(f"{ratio:4}%  ")
            
            # Compare files
            if not self.files_are_equal(filename, output_file):
                self.log_file.write("Failed\n")
                self.total_failed += 1
                return False
            
            self.log_file.write("Passed\n")
            self.total_passed += 1
            return True
            
        except subprocess.TimeoutExpired:
            self.log_file.write("Failed: Command timed out\n")
            self.total_failed += 1
            return False
        except Exception as e:
            self.total_failed += 1
            self.log_file.write(f"Failed: {e}\n")
            return False
            
        finally:
            # Clean up temporary files
            for temp_file in [compressed_file, output_file]:
                if os.path.exists(temp_file):
                    try:
                        os.remove(temp_file)
                    except:
                        pass
    
    def files_are_equal(self, file1, file2):
        if os.path.getsize(file1) != os.path.getsize(file2):
            return False
        
        buffer_size = 4096
        with open(file1, 'rb') as f1, open(file2, 'rb') as f2:
            while True:
                chunk1 = f1.read(buffer_size)
                chunk2 = f2.read(buffer_size)
                
                if not chunk1 and not chunk2:
                    break
                    
                if chunk1 != chunk2:
                    return False
        
        return True
    
    def write_log_header(self):
        """Write header to log file"""
        self.log_file.write("                                          Original   Packed\n")
        self.log_file.write("            File Name                     Size      Size   Ratio  Result\n")
        self.log_file.write("-------------------------------------     --------  --------  ----  ------\n")
    
    def write_log_summary(self, start_time, stop_time):
        """Write summary to log file"""
        elapsed = stop_time - start_time
        self.log_file.write(f"\nTotal elapsed time: {elapsed:.2f} seconds\n")
        self.log_file.write(f"Total files:   {self.total_files}\n")
        self.log_file.write(f"Total passed:  {self.total_passed}\n")
        self.log_file.write(f"Total failed:  {self.total_failed}\n")
    
    def usage_exit(self):
        """Display usage information and exit"""
        usage = """
CHURN 1.0. Usage: CHURN root-dir "compress command" "expand command"

CHURN tests compression programs by compressing and expanding all files in a directory.

Examples:
  Windows:
   python churn.py C:\\Dir "Lzss-c.exe %s TEST.CMP" "Lzss-e.exe -d TEST.CMP TEST.OUT"
  
  Linux:
    python churn.py /home/user "gzip -c %s TEST.CMP" "gzip -d TEST.CMP TEST.OUT"
    
    python churn.py /home/user "gzip -c %s" "gzip -d -c TEST.CMP"
    
    python churn.py /any/RootDir "./Lzss-c %s TEST.CMP" "./Lzss-e TEST.CMP TEST.OUT"

Note: Use %s as placeholder for input filename in compress command.
      TEST.CMP and TEST.OUT will be automatically created in the current directory.
"""
        print(usage)
        sys.exit(1)


if __name__ == "__main__":
    churn = ChurnProgram()
    churn.main(sys.argv[1:])
